src/lib.rs

```diff
@@ -392,58 +460,31 @@ impl KafkaJsonToDelta {
 
-            // Lookup the last transaction version from the Delta Log
-            let last_txn_version = state
-                .delta_writer
-                .last_transaction_version(self.app_id.as_str())
-                .await?;
-
-            // Prepare the new WAL entry
-            let wal_timer = Instant::now();
-            let wal_entry = if let Some(txn_version) = last_txn_version {
-                self.log_delta_tx_version_found(self.app_id.as_str(), txn_version)
-                    .await;
-
-                let last_wal_entry = state.wal.get_entry_by_transaction_id(txn_version).await?;
-
-                last_wal_entry.prepare_next(&partition_offsets)
-            } else {
-                self.log_delta_tx_version_not_found(self.app_id.as_str())
-                    .await;
-                WriteAheadLogEntry::new(
-                    1,
-                    TransactionState::Prepared,
-                    None, /*delta table version starts out empty*/
-                    partition_offsets.clone(),
-                )
-            };
-            state.wal.put_entry(&wal_entry).await?;
-            self.log_write_ahead_log_prepared().await;
-
             // Complete the delta write
             let delta_write_timer = Instant::now();
-            let txn = action::Action::txn(action::Txn {
-                appId: self.app_id.clone(),
-                version: wal_entry.transaction_id,
-                lastUpdated: std::time::SystemTime::now()
-                    .duration_since(std::time::UNIX_EPOCH)
-                    .unwrap()
-                    .as_millis() as i64,
-            });
+
+            let mut txns = Vec::new();
+            for (partition, offset) in partition_offsets {
+                let txn = action::Action::txn(action::Txn {
+                    appId: self.app_id_for_partition(partition),
+                    version: offset,
+                    lastUpdated: std::time::SystemTime::now()
+                        .duration_since(std::time::UNIX_EPOCH)
+                        .unwrap()
+                        .as_millis() as i64,
+                });
+
+                txns.push(txn);
+            }
+
             self.log_delta_write_started().await;
-            let commit_result = state.delta_writer.write_file(&mut vec![txn]).await;
+            let commit_result = state.delta_writer.write_file(&mut txns).await;
             match commit_result {
                 Ok(version) => {
-                    self.log_delta_write_completed(&delta_write_timer).await;
-                    state
-                        .wal
-                        .complete_entry(wal_entry.transaction_id, version)
-                        .await?;
-                    self.log_write_ahead_log_completed(&wal_timer).await;
+                    self.log_delta_write_completed(version, &delta_write_timer)
+                        .await;
                 }
                 Err(e) => {
```

Comment:
please let me know if I read this correctly or not. in case of concurrent commit conflicts, i.e. 2 workers committing to the same version, the other worker won, we will get a commit error here. It will propagate the error all the way back to the run_loop and crash the application with an error?

Reply:
That's correct at this time. With this PR @mosyp is removing the write ahead log, replacing with a txn per partition in the delta log and adding tests for concurrent workers. We will follow-up immediately afterwards with an optimistic concurrency loop that knows how to do txn conflict resolution in case of rebalance. I've added https://github.com/delta-io/kafka-delta-ingest/issues/25 to make this separation of work clear.

Reply:
sounds good :+1: just wanted to make sure the conflict resolution is planned so we are restarting most of the workers all the time :)

Reply:
We're planning to use our own optimistic concurrency loop but this
> 2 workers committing to the same version, the other worker won, we will get a commit error here

should not be the case, since we're doing commit through delta-rs's `try_commit_loop` which should handle concurrency

Reply:
but there's should be another issue that requires our own `try_commit_loop`, let me do some digging

Reply:
`try_commit_loop` can only automatically resolve conflicts for commits that only contains add file action right? what happens when there is a conflict on offsets caused by rebalance?

Reply:
yes, this part with offsets I was missing. I'm going to comment out the test with validating a data until conflict resolution is implemented